providers/provider-types.go

```diff
@@ -0,0 +1,63 @@
+package providers
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"strings"
+	"sync"
+)
+
+type NodeProvider interface {
+	ListNodes(reader *bufio.Reader) ([]string, error)
+	ReadNodeConfigurations(deployNodes []string) ([]v3.RKEConfigNode, error)
```

Comment:
Lets add Register to the list of methods. Things like setting up environment variables have be done at registration point, and it should fail if something goes wrong

Reply:
You can look at these code bases as an example of providers registration: https://github.com/rancher/lb-controller/blob/master/provider/provider.go or https://github.com/rancher/external-dns/blob/master/providers/provider.go